hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -19,29 +38,232 @@ impl TraceDecoder for YkPTTraceDecoder {
+    MainObj(u64),
+    /// Either a location not in the "main binary object", or an unknown location.
+    OtherObjOrUnknown,
+}
+
+impl Debug for ObjLoc {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::MainObj(v) => write!(f, "MainObj[0x{:x}]", v),
+            Self::OtherObjOrUnknown => write!(f, "OtherObjOrUnknown"),
+        }
+    }
+}
+
 /// Iterate over the blocks of an Intel PT trace using the fast Yk PT decoder.
 struct YkPTBlockIterator<'t> {
     /// Set to true when an error has occured.
     errored: bool,
-    /// PT packet iterator.
+    /// The packet iterator used to drive the decoding process.
     parser: PacketParser<'t>,
+    /// Keeps track of where we are in the traced binary.
+    cur_loc: ObjLoc,
+    /// A vector of "taken/not-taken" (TNT) decisions. These come in in batches and get buffered
+    /// here in a FIFO fashion (oldest decision at head poistion).
+    tnts: VecDeque<bool>,
+    /// When true, packet generation is enabled (we've seen a `TIP.PGE` packet, but no
+    /// corresponding `TIP.PGD` yet).
+    pge: bool,
+}
+
+impl<'t> YkPTBlockIterator<'t> {
+    /// Looks up the blockmap entry for the given offset in the "main object binary".
+    fn lookup_blockmap_entry(
+        off: u64,
+    ) -> Option<&'static intervaltree::Element<u64, BlockMapEntry>> {
+        let mut ents = LLVM_BLOCK_MAP.query(off, off + 1);
+        if let Some(ent) = ents.next() {
+            // A single-address range cannot span multiple blocks.
+            debug_assert!(ents.next().is_none());
+            Some(ent)
+        } else {
+            None
+        }
+    }
+
+    // Lookup a block from an offset in the "main binary" (i.e. not from a shared object).
+    fn lookup_block_from_main_bin_offset(&self, off: u64) -> Block {
+        if let Some(ent) = Self::lookup_blockmap_entry(off) {
+            Block::from_vaddr_range(
+                u64::try_from(off_to_vaddr(&PathBuf::from(""), ent.range.start).unwrap()).unwrap(),
+                u64::try_from(off_to_vaddr(&PathBuf::from(""), ent.range.end).unwrap()).unwrap(),
+            )
+        } else {
+            Block::unknown()
+        }
+    }
+
+    fn do_next(&mut self) -> Result<Option<Block>, HWTracerError> {
+        // Read as far ahead as we can using static successor info encoded into the blockmap.
+        if let ObjLoc::MainObj(b_off) = self.cur_loc {
+            // We know where we are in the main object binary, so there's a chance that there's a
+            // blockmap entry for this location.
+            let ent = Self::lookup_blockmap_entry(b_off.to_owned());
+            if ent.is_none() {
+                // In the absence of blockmap info, we cannot know the successor block.
+                self.cur_loc = ObjLoc::OtherObjOrUnknown;
+                return Ok(Some(Block::unknown()));
+            }
+            let ent = ent.unwrap();
```

Comment:
Even though this `unwrap` is safe now, we should avoid these whenever possible. In this case, I think we can not only do that, but make the code clearer with:

```rust
let ent = match self.lookup_blockmap_entry(...) {
  Some(x) => x,
  None => {
    // In the absence of blah blah
    ...
  }
}
```

Reply:
In this case I disagree. I dislike introducing long indented scopes. I find it really gets in the way of comprehension.

Reply:
Fixed in 3f6535d0c5dd64637ac2e9f40c0c968a64e05170.